src/messaging/ExchangeHolder.h

```diff
@@ -0,0 +1,139 @@
+
+    /**
+     * @brief
+     *
+     * Replaces the held exchange and associated delegate to instead track the given ExchangeContext, aborting
+     * and dereferencing any previously held exchange as necessary. This method should be called whenever protocol logic
+     * that is managing this holder is transitioning from an outdated Exchange to a new one, often during
+     * the start of a new transaction.
+     */
+    void Grab(ExchangeContext * exchange)
+    {
+        Release();
+
+        mpExchangeCtx = exchange;
+        mpExchangeCtx->SetDelegate(this);
+    }
+
+    /*
+     * @brief
+     *
+     * This shuts down the exchange (if a valid one is being tracked) and releases our reference to it.
+     */
+    void Release()
+    {
+        if (mpExchangeCtx)
+        {
+            mpExchangeCtx->SetDelegate(nullptr);
+
+            /**
+             * Shutting down the exchange requires calling Abort() on the exchange selectively in the following scenarios:
+             *      1. The exchange is currently awaiting a response. This would have happened if our consumer just sent a message
+             * on the exchange and is awaiting a response. Since we no longer have an interest in this exchange anymore, we should
+             * abort it to release our reference.
+             *
+             *      2. Our consumer has signaled an interest in sending a message. This could have been signaled right at exchange
+             * creation time as the initiator, or when handling a message and the consumer intends to send a response, albeit,
+             * asynchronously. In both cases, the stack expects the exchange consumer to close/abort the EC if it no longer has
+             * interest in it.
+             */
+            if (mpExchangeCtx->IsResponseExpected() || mpExchangeCtx->IsSendExpected())
+            {
+                mpExchangeCtx->Abort();
+            }
+        }
+
+        mpExchangeCtx = nullptr;
+    }
+
+    explicit operator bool() const { return mpExchangeCtx != nullptr; }
+    ExchangeContext * Get() const { return mpExchangeCtx; }
+
+    ExchangeContext * operator->() const
+    {
+        VerifyOrDie(mpExchangeCtx != nullptr);
+        return mpExchangeCtx;
+    }
+
+private:
+    CHIP_ERROR OnMessageReceived(ExchangeContext * ec, const PayloadHeader & payloadHeader,
+                                 System::PacketBufferHandle && payload) override
+    {
+        return mpExchangeDelegate.OnMessageReceived(ec, payloadHeader, std::move(payload));
+    }
+
+    void OnResponseTimeout(ExchangeContext * ec) override { return mpExchangeDelegate.OnResponseTimeout(ec); }
+
+    void OnExchangeClosing(ExchangeContext * ec) override
+    {
+        if (mpExchangeCtx)
+        {
+            mpExchangeCtx->SetDelegate(nullptr);
+            mpExchangeCtx = nullptr;
+        }
+
+        mpExchangeDelegate.OnExchangeClosing(ec);
+    }
```

Comment:
Don't we need to forward `GetMessageDispatch` too?

Reply:
Missed that. Will add.